core/src/main/java/org/infinispan/metadata/EmbeddedMetadata.java

```diff
@@ -235,21 +233,29 @@ public String toString() {
 
    public static class EmbeddedLifespanExpirableMetadata extends EmbeddedMetadata {
 
-      @ProtoField(number = 3, defaultValue = "-1")
-      public long lifespan;
-
-      protected EmbeddedLifespanExpirableMetadata() {}
+      final long lifespan;
 
-      protected EmbeddedLifespanExpirableMetadata(long lifespan, TimeUnit lifespanUnit, EntryVersion version) {
+      protected EmbeddedLifespanExpirableMetadata(long lifespan, EntryVersion version) {
          super(version);
-         this.lifespan = lifespan < 0 ? -1 : lifespanUnit.toMillis(lifespan);
+         this.lifespan = lifespan;
+      }
+
+      @ProtoFactory
+      protected EmbeddedLifespanExpirableMetadata(long lifespan, NumericVersion numericVersion, SimpleClusteredVersion clusteredVersion) {
+         this(lifespan, numericVersion != null ? numericVersion : clusteredVersion);
       }
 
       @Override
       public long lifespan() {
          return lifespan;
       }
 
+      // Need a separate getter method as needs to conform to protostream getter format and has to be accessed via MemcachedMetadata
```

Comment:
The annotation can be moved to the field and the field should be marked protected (because will be used by a subclass).

Reply:
Oh, I wrote that without thinking. The generated marshaller is not a subclass ... so it cannot see the protected field :)